### DISCUSSION TITLE
Biigle served on subdirectory of domain?

### DISCUSSION THREAD

**Discussion by cywhale** [biigle/core#340] [Q&A] [answered]
I want to serve Biigle on subdirectory of domain, i.e. https://mydomain/biigle/, for example.
By using proxy_pass in NGINX or rewrite rules, such as

```
upstream biigle {
  server 127.0.0.1:8008;
}

location \biigle {
  rewrite ^/biigle(.*)$ /$1 break;
  proxy_pass https://biigle;
}

location ~ ^/(assets|vendor|storage|doc)($|/) {
    proxy_pass https://biigle;
}
```
I can get initial layouts. But if internally routes to login/logout/search ..., no surprise got 404.
Is there any other way can do this? Thanks.

**Comment by mzur** [comment, marked as answer]
I'm not an expert in NGINX as reverse proxy but shouldn't this be enough?

```nginx
upstream biigle {
  server 127.0.0.1:8008;
}

location /biigle/ {
  proxy_pass https://biigle/;
}
```
You shouldn't need a rewrite or an extra location for `assets|vendor|...` as this is handled by the webserver of the BIIGLE Docker container.

You also need to configure the [`APP_URL`](https://github.com/biigle/biigle/blob/6febd521ff0cd96ee895da27f72640f02ab2ae8e/build/.env.example#L9) to the correct URL (with subdirectory) but this does not affect regular links in the application.

**Reply by cywhale** [reply]
I had tried that. First, modify "APP_URL" to "https://mydomain/biigle" has no effect to this issue. Seems biigle/core internally use domain or host only.  If use upstream only (like you mentioned above), the static assets cannot be fetched and then got 404 like this:

![https://github.com/cywhale/ODB/blob/master/ROV/biigle/clip.png](https://raw.githubusercontent.com/cywhale/ODB/master/ROV/biigle/clip.png)

If append rules:
```
location ~ ^/(assets|vendor|storage|doc)($|/) {
    proxy_pass https://biigle;
}
```
the static assets then can be fetched, but cannot serve mydomain/biigle/index.php well, just like

![https://github.com/cywhale/ODB/blob/master/ROV/biigle/clip1.png](https://raw.githubusercontent.com/cywhale/ODB/master/ROV/biigle/clip1.png)

If append rewrite rules and already logined, I can get initial layouts well:
```
rewrite ^/biigle(.*)$ /$1 break;
```

![https://github.com/cywhale/ODB/blob/master/ROV/biigle/clip2.png](https://raw.githubusercontent.com/cywhale/ODB/master/ROV/biigle/clip2.png)

But after sign out, logout/login, or perform search/api, those functions that would do internal routing, everything goes 404 error again.  I still try hard for this, and wonder if there is any alternatives on this issue?~~

**Reply by mzur** [reply]
I see. My proxy configuration above should still be enough. But BIIGLE needs [additional configuration](https://laravel.com/docs/6.x/requests#configuring-trusted-proxies) for the reverse proxy. Try this:

1. Create a new file `trustedproxy.php` in your build setup [here](https://github.com/biigle/biigle/tree/6febd521ff0cd96ee895da27f72640f02ab2ae8e/build/config). The file should have the content:
   ```php
   <?php

   return [
      'proxies' => [
         '127.0.0.1',
      ],
   ];
   ```
2. Add a new `COPY` statement for the config file to the [dockerfile](https://github.com/biigle/biigle/blob/6febd521ff0cd96ee895da27f72640f02ab2ae8e/build/build.dockerfile#L68).
3. Rebuild the Docker containers and restart BIIGLE.

The NGINX reverse proxy adds new `X-Forwarded*`  HTTP headers to the requests as described [here](https://www.nginx.com/resources/wiki/start/topics/examples/forwarded/). These should tell BIIGLE what URLs to use. But it only does so if the headers come from the IP of a "trusted" proxy (else anyone could fake these headers). In this case BIIGLE should trust your proxy that runs on the same machine.

There is one caveat though. It might be that your proxy has a different IP than `127.0.0.1` because of the networking of the Docker containers. You have to find out the correct host IP then.

**Reply by cywhale** [reply]
Get some progress from your response. 
The modifications I did:
1. modify APP_URL: "https://mydomain/biigle" <-- but seems not affect at all
2. add trustedproxy.php to docker build, and add my host IP, docker IP, 127.0.0.1 to proxies
3. modify nginx upstream as your configuration, and add
```
upstream biigle {
  server 127.0.0.1:8008;
}

location /biigle/ {
  proxy_pass https://biigle/;

  proxy_set_header    X-Real-IP           $remote_addr;
  proxy_set_header    X-Forwarded-For     $proxy_add_x_forwarded_for;
  proxy_set_header    X-Forwarded-Proto   $scheme;
  proxy_set_header    Host                $host; ## <--- I think this is only "effective", modify it will got total 404. Give it $host can get initial layout well. I marked other X-Forwarded header or modify as other values seems no change to the results. 
  proxy_set_header    X-Forwarded-Host    $host;
  proxy_set_header    X-Forwarded-Port    $server_port;
}
```

But I still need this snippet to get correct layout: 
```
 location ~ ^/(assets|vendor|storage|doc|fonts)($|/) {
    proxy_pass https://biigle;
 } 
```
![https://github.com/cywhale/ODB/blob/master/ROV/biigle/clip3.png](https://raw.githubusercontent.com/cywhale/ODB/master/ROV/biigle/clip3.png)

But when click any link on Biigle, it will still internally route to https://mydomain/, thus got 404 like this: (For example, I click "Create project" here). Window location will be changed internally to 'https//mydomain/routing' (i.e, no /biigle/ ). You can see the address bar on browser.

![https://github.com/cywhale/ODB/blob/master/ROV/biigle/clip4.png](https://raw.githubusercontent.com/cywhale/ODB/master/ROV/biigle/clip4.png)

If manually keyin "/biigle/" in address bar on browser, then get correct layout back:

![https://github.com/cywhale/ODB/blob/master/ROV/biigle/clip5.png](https://raw.githubusercontent.com/cywhale/ODB/master/ROV/biigle/clip5.png)

So the problem is still that biigle/core use 'https://mydomain' only internally no matter I modify APP_URL to 'https://mydomain/biigle' or change X-Forward* header to 'https://mydomain/biigle' (But "proxy_set_header    Host                $host;" cannot be modified, otherwise cause totally 404 error).

**Reply by mzur** [reply]
My bad, the TrustProxies middleware is [disabled](https://github.com/biigle/core/blob/2547db96819affd90bcc66ddb965b62dd6b00a57/app/Http/Kernel.php#L17) by default. Try to add this to the dockerfile:

```dockerfile
RUN sed -i 's/\/\/ \\Biigle/\\Biigle/' app/Http/Kernel.php
```

**Reply by cywhale** [reply]
Thanks. I can confirm that app/Http/Kernel.php and trustedproxy.php have been modified/added following your suggestions by checking these files in docker container. But the results remain the same: I can get only initial layout well, but when clicking other links, the window location push https://mydomain/(routing) and then get 404.

I copied a test.include into /etc/nginx/conf.d/ by add COPY in web.dockerfile to debug:
```
    add_header x-uri "$uri" always;
    add_header x-NGINX-request_uri $request_uri always;
    add_header x-host $host always;
    add_header http-host $http_host always;
    add_header X-Real-IP $remote_addr always;
    add_header X-Forwarded-For $proxy_add_x_forwarded_for always;

```
I get confused about different proxy_set_header configuration would get weird response header:
1. When set  proxy_set_header  Host $host;  proxy_set_header X-Forwarded-Host $host;
==>  http_host = x_host = mydomain,  and x-uri is /index.php. The result is described as aboving (only initial is good):

![https://github.com/cywhale/ODB/blob/master/ROV/biigle/clip8.png](https://raw.githubusercontent.com/cywhale/ODB/master/ROV/biigle/clip8.png)
 
2. When set proxy_set_header  Host "mydomain/biigle"
==> http_host = "mydomain/biigle",  but x-host becomes "example.com",  x-uri becomes "/"  (NOT /index.php). 
So this setting "mydomain/biigle" not work and got 400 error.

![https://github.com/cywhale/ODB/blob/master/ROV/biigle/clip6.png](https://raw.githubusercontent.com/cywhale/ODB/master/ROV/biigle/clip6.png)

**Reply by mzur** [reply]
It turns out that running a Laravel application in a subfolder is quite hard. Laravel is not made for this. I created an [experimental setup](https://github.com/biigle/biigle/compare/revproxy-subfolder) with a reverse proxy and a subfolder. It works somewhat (except the initial login redirect, and the logo and icon font which are loaded via CSS). New are the [`ASSET_URL`](https://github.com/biigle/biigle/compare/revproxy-subfolder#diff-8d92e514dbcf8eb8215e7c4eaac59c33461fc9a043b384c02b78c5e532ac9486R10) and a [`forceRootUrl`](https://github.com/biigle/biigle/compare/revproxy-subfolder#diff-d63ad7108957a8e7fde88ddf92c9715229f68f871796cf1d4e0802f56ca736bdR73) hack.

I suggest you use a subdomain instead of a subdirectory.

**Reply by cywhale** [reply]
Yeah, I'll try and at least I had a working Biigle setting on ' \ '. It use a subdomain, indeed, and for our limited IP/server resources, so that I consider move it to subdirectory by upstreaming. Anyway, you really give me a big hand, thanks a lot.

**Reply by mzur** [reply]
You're welcome. And last time I checked, NGINX could handle several domains (even with SSL) so you can use the same IP and server for multiple subdomains.
